### PR TITLE
Introduce expression rewriter

### PR DESCRIPTION
**Pull request by FabioLuporini**
** Premise ** I'm filing this PR so that people can start looking into it, *but* it still lacks a comprehensive performance evaluation. This will hopefully be integrated soon.

This PR extends the symbolic engine in Devito. Besides a cleaner interface, the most important addition is the ``factorizer``, a rewrite operator which gets applied right after CSE. Basically, using the ``collect_nested`` function (a contribution of this PR, since sympy's built-ins just suck for our purposes), constants and common temporaries are sought in each expression composing the stencil (ie the stencils themselves plus any sub-expression captured by CSE) and factorized.

These are the flop reductions I got.
Problem: acoustic; each line is a different space order ([2->16, step=2]).
```
Rewriter: 28 --> 23 flops (Gain: 1.22 X)
Rewriter: 40 --> 30 flops (Gain: 1.33 X)
Rewriter: 52 --> 37 flops (Gain: 1.41 X)
Rewriter: 64 --> 44 flops (Gain: 1.45 X)
Rewriter: 76 --> 51 flops (Gain: 1.49 X)
Rewriter: 88 --> 58 flops (Gain: 1.52 X)
Rewriter: 100 --> 65 flops (Gain: 1.54 X)
Rewriter: 112 --> 72 flops (Gain: 1.56 X)
```

Problem: TTI; each line is a different space order ([2->6, step=2]).
```
Rewriter: 389 --> 327 flops (Gain: 1.19 X)
Rewriter: 935 --> 839 flops (Gain: 1.11 X)
Rewriter: 1783 --> 1667 flops (Gain: 1.07 X)
``` 

The gain dropping with TTI is due to CSE, which "hides" some factorisation opportunities. A way to work around this would consist of searching "across" different expressions and somehow rearranging the temporaries. This requires further thinking and some legwork, so it's not gonna be shipped with this PR.

Some new tests are introduced.

TODO:
- Run experiments to collect runtimes and understand how the roofline plots change

### CODE REVIEW

**Comment by FabioLuporini** [comment]
I deliberately left the "DROP BEFORE MERGE" commits for the moment so that interested people can quickly switch to this branch and try out the new functionality by simply running, for example, ``problem=acoustic order=2 grid=64 mode=original ./launcher.sh``, from the examples folder

**Comment by FabioLuporini** [comment]
I deliberately left the "DROP BEFORE MERGE" commits for the moment so that interested people can quickly switch to this branch and try out the new functionality by simply running, for example, ``problem=acoustic order=2 grid=64 mode=original ./launcher.sh``, from the examples folder

File: devito/symbolics.py [hunk 1]
```diff
@@ -83,90 +90,190 @@ def dse_indexify(expr):
...
-
-    :return: A list of the resulting equations after performing CSE
+    Transform expressions to create time-invariant computation.
     """
-    expr = expr if isinstance(expr, list) else [expr]
 
-    temps, stencils = cse(expr, numbered_symbols("temp"))
+    if mode == True:
+        return Rewriter(expr).run(mode='advanced')
+    elif mode in ['basic', 'advanced']:
```
Comment (mlange05) [inline devito/symbolics.py]: These modes don't seem to be demonstrated anywhere?

File: devito/compiler.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def __init__(self, *args, **kwargs):
         super(IntelCompiler, self).__init__(*args, **kwargs)
         self.cc = 'icpc'
         self.ld = 'icpc'
-        self.cflags = ['-O3', '-g', '-fPIC', '-Wall', "-xhost"]
+        self.cflags = ['-O2', '-g', '-fPIC', '-Wall', "-xhost"]
```
Comment (FabioLuporini) [inline devito/compiler.py]: Oh no, this is part of the "DROP BEFORE MERGE" commits

File: devito/symbolics.py (hunk 1 above)
Comment (FabioLuporini) [inline devito/symbolics.py]: True. I'll add a couple of lines. Basically:
basic -> CSE
advanced -> CSE + Factorization

We default to ``advanced`` for obvious reasons we want to have both
Will fix this !

**Comment by FabioLuporini** [comment]
update:
doc fixes in
flake8 fixes in

**Comment by FabioLuporini** [comment]
@mlange05 can we update the following requirement:
``sympy>=0.7.6`` ---> ``sympy>=1.0.0`` ?

**Pull request closed without merging**